core/src/test/java/org/bitcoinj/wallet/WalletTest.java

```diff
@@ -3441,4 +3442,39 @@ public void totalReceivedSent() throws Exception {
 
         // TODO: test shared wallet calculation here
     }
+
+    @Test
+    public void testIrrelevantDoubleSpend() throws Exception {
+        // 1 spends 0
+        Transaction tx0 = createFakeTx(PARAMS, COIN.multiply(5), OTHER_ADDRESS, OTHER_ADDRESS2)[1];
+        Transaction tx1 = createFakeTx(PARAMS, COIN.multiply(5), OTHER_ADDRESS, OTHER_ADDRESS2)[1];
+
+        Transaction tx2 = new Transaction(PARAMS);
+        tx2.addInput(tx0.getOutput(0));
+        tx2.addOutput(COIN, myAddress);
+        tx2.addOutput(COIN, OTHER_ADDRESS);
+
+        sendMoneyToWallet(AbstractBlockChain.NewBlockType.BEST_CHAIN, tx2, tx1, tx0);
+
+        // tx3 and tx4 double spend each other
+        Transaction tx3 = new Transaction(PARAMS);
+        tx3.addInput(tx1.getOutput(0));
+        tx3.addOutput(COIN, myAddress);
+        tx3.addOutput(COIN, OTHER_ADDRESS);
+        wallet.receivePending(tx3, null);
+
+        // tx4 also spends irrelevant output from tx2
+        Transaction tx4 = new Transaction(PARAMS);
+        tx4.addInput(tx1.getOutput(0)); // spends same output
+        tx4.addInput(tx2.getOutput(1));
+        tx4.addOutput(COIN, OTHER_ADDRESS);
+
+        // tx4 does not actually get added to wallet here since it by itself is irrelevant
+        sendMoneyToWallet(AbstractBlockChain.NewBlockType.BEST_CHAIN, tx4);
+
+        // since tx4 is not saved, tx2 output 1 will have bad spentBy
+        wallet = roundTrip(wallet);
+
+        log.info(wallet.toString()); // will throw exception if state invalid
```

Comment:
We shouldn't log.info() this, it will spam the log.

Actually we should be able to use Wallet.isConsistent() here. If isConsistent currently isn't sensitive to this problem, it should be changed so it detects such issues.